File: app/src/main/java/com/paradoxo/amadeus/service/GravaHistoricoService.java
package com.paradoxo.amadeus.service;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;

import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.paradoxo.amadeus.dao.SentencaDAO;
import com.paradoxo.amadeus.modelo.Sentenca;
import com.paradoxo.amadeus.util.Preferencias;

import java.util.List;

public class GravaHistoricoService extends Service {
    public static final String TK = "tk";
    public static final String ID_ULTIMO_UPLOAD = "idUltimoUpload";

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        long idUltimoUpload = intent.getLongExtra(ID_ULTIMO_UPLOAD, -1);
        Log.e(ID_ULTIMO_UPLOAD, String.valueOf(idUltimoUpload));

        SentencaDAO sentencaDAO = new SentencaDAO(this, true);
        List<Sentenca> historicoParaUpload = sentencaDAO.listarAPartirDe(idUltimoUpload);
        inseriHistoricoFibrebase(historicoParaUpload);

        return super.onStartCommand(intent, flags, startId);
    }

    private void inseriHistoricoFibrebase(List<Sentenca> historicoParaUpload){
        String tk = Preferencias.getPrefString(TK, getApplicationContext());

        DatabaseReference databaseReference = FirebaseDatabase.getInstance().getReference();
        databaseReference.child("historico-beta").child(tk).push().setValue(historicoParaUpload)
                .addOnSuccessListener(aVoid -> Preferencias.setPrefLong(ID_ULTIMO_UPLOAD,
                        Long.parseLong(historicoParaUpload.get(historicoParaUpload.size()-1).getId()),
                        getApplicationContext())).addOnFailureListener(e -> {
                    Log.e("Falha ao upar", "Histórico de conversas");
                    e.printStackTrace();
                });

        this.stopSelf();
        // Depois que termina de fazer o upload para a service
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.e("Service Historico", "Destruida");
    }

}

File: app/src/main/java/com/paradoxo/amadeus/service/GravaHistoricoService.kt
package com.paradoxo.amadeus.service

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.util.Log
import com.google.firebase.database.FirebaseDatabase
import com.paradoxo.amadeus.dao.room.AmadeusDatabase
import com.paradoxo.amadeus.dao.room.toModel
import com.paradoxo.amadeus.modelo.Sentenca
import com.paradoxo.amadeus.util.Preferencias
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class GravaHistoricoService : Service() {
    companion object {
        const val TK = "tk"
        const val ID_ULTIMO_UPLOAD = "idUltimoUpload"
    }

    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val idUltimoUpload = intent.getLongExtra(ID_ULTIMO_UPLOAD, -1)
        Log.e(ID_ULTIMO_UPLOAD, idUltimoUpload.toString())

        serviceScope.launch {
            val historico = AmadeusDatabase.getInstance(this@GravaHistoricoService)
                .sentencaDAO().listarAPartirDe(idUltimoUpload)
            if (historico.isNotEmpty()) {
                inseriHistoricoFibrebase(historico.map { it.toModel() })
            } else {
                stopSelf()
            }
        }

        return super.onStartCommand(intent, flags, startId)
    }

    private fun inseriHistoricoFibrebase(historicoParaUpload: List<Sentenca>) {
        val tk = Preferencias.getPrefString(TK, applicationContext)

        FirebaseDatabase.getInstance().reference
            .child("historico-beta").child(tk).push()
            .setValue(historicoParaUpload)
            .addOnSuccessListener {
                Preferencias.setPrefLong(
                    ID_ULTIMO_UPLOAD,
                    historicoParaUpload.last().id!!.toLong(),
                    applicationContext
                )
            }
            .addOnFailureListener { e ->
                Log.e("Falha ao upar", "Histórico de conversas")
                e.printStackTrace()
            }

        stopSelf()
    }

    override fun onBind(intent: Intent): IBinder? = null

    override fun onDestroy() {
        super.onDestroy()
        serviceScope.cancel()
        Log.e("Service Historico", "Destruida")
    }
}
